File: golang-iac/src/util/eks-config/eks_cluster_config.go
package eksconfig

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"strings"

	"github.com/aws/aws-sdk-go-v2/aws"

	CFN "aws/cfn"
	EC2 "aws/ec2"
	SECRETS "aws/secretsmanager"
)

func ConfigureEks(cfg aws.Config, stackName string, p map[string]string) {

	// get eks cluster name (same name as cfn stack name)
	eksClusterName := CFN.GetNestedStackName(cfg, stackName, "Eks")
	vpcStackName := CFN.GetNestedStackName(cfg, stackName, "Vpc")
	vpcStackOutput := CFN.GetStackOutputs(cfg, vpcStackName)

	for _, i := range strings.Split(vpcStackOutput["PublicSubnetGroup"], ",") {
		EC2.CreateTags(cfg, i, "kubernetes.io/cluster/"+eksClusterName, "shared")
		EC2.CreateTags(cfg, i, "kubernetes.io/role/elb", "1")
	}

	for _, i := range strings.Split(vpcStackOutput["PrivateSubnetGroup"], ",") {
		EC2.CreateTags(cfg, i, "kubernetes.io/cluster/"+eksClusterName, "shared")
		EC2.CreateTags(cfg, i, "kubernetes.io/role/internal-elb", "1")
	}

	// add permissions to allow users access to eks cluster
	eksStackOutput := CFN.GetStackOutputs(cfg, eksClusterName)

	// based on evironment name, select the dev role to allow access to eks to engineering
	roleArn := map[string]string{
		"dev":     "arn:aws:iam::597609335369:role/CrossAccountDevRole",
		"testing": "arn:aws:iam::597609335369:role/CrossAccountDevRole",
		"qa":      "arn:aws:iam::667813852149:role/CrossAccountDevRole",
		"test":    "arn:aws:iam::667813852149:role/CrossAccountDevRole",
		"staging": "arn:aws:iam::667813852149:role/CrossAccountDevRole",
		"main":    "arn:aws:iam::394007061642:role/CrossAccountDevRole",
	}

	var role string
	if roleArn[p["Name"]] == "" {
		//default to dev if this is an edge environment name
		role = "arn:aws:iam::597609335369:role/CrossAccountDevRole"
	} else {
		role = roleArn[p["Name"]]
	}

	data := []byte(`
apiVersion: v1
kind: ConfigMap
metadata:
  name: aws-auth
  namespace: kube-system
data:
  mapRoles: |
    - rolearn: ` + eksStackOutput["EksNodeInstanceRole"] + `
      username: system:node:{{EC2PrivateDNSName}}
      groups:
        - system:bootstrappers
        - system:nodes
    - rolearn: ` + role + `
      username: SAMladmin
      groups:
        - system:masters
        - cluster-admin
  mapUsers: |
    - userarn: ` + eksStackOutput["EKSUserArn"] + `
      username: ` + eksStackOutput["EKSUserName"] + `
      groups:
        - system:masters
        - cluster-admin
    `)

	// check if configmap aws-auth already exists
	// not sure if it's needed

	err := os.WriteFile("/tmp/aws-auth.yaml", data, 0644)
	if err != nil {
		log.Println(err)
	}

	cmd := exec.Command("kubectl", "apply", "-f", "/tmp/aws-auth.yaml")
	//response, err := cmd.Output()
	var out bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &stderr
	err = cmd.Run()

	if err != nil {
		log.Println(fmt.Sprint(err) + ": " + stderr.String())
		//log.Println(err)
	} else {
		log.Println("Applied kube template: " + out.String())
	}

	// deploy metrics server
	cmd = exec.Command("kubectl", "apply", "-f", "https://github.com/kubernetes-sigs/metrics-server/releases/latest/download/components.yaml")
	response, err := cmd.Output()
	if err != nil {
		log.Println(err)
	} else {
		log.Println("Deployed metrics server: " + string(response))
	}

	// deploy kubernetes dashboard
	cmd = exec.Command("kubectl", "apply", "-f", "https://raw.githubusercontent.com/kubernetes/dashboard/v2.7.0/aio/deploy/recommended.yaml")
	response, err = cmd.Output()
	if err != nil {
		log.Println(err)
	} else {
		log.Println("Deployed kubernetes dashboard: " + string(response))
	}

	// Install the nvidia plugin
	cmd = exec.Command("kubectl", "apply", "-f", "../conf/kube/nvidia-plugin.yaml")
	response, err = cmd.Output()
	if err != nil {
		log.Println(err)
	} else {
		log.Println("Deployed nvidia plugin: " + string(response))
	}

	// Cluster AutoScaler file path per EKS cluster
	input, err := ioutil.ReadFile("../conf/kube/cluster-autoscaler-autodiscover.yaml")
	if err != nil {
		log.Println(err)
	}

	output := bytes.Replace(input, []byte("<YOUR_CLUSTER_NAME>"), []byte("something"), -1)

	err = os.WriteFile("/tmp/autoscaling.yaml", output, 0644)
	if err != nil {
		log.Println(err)
	}

	cmd = exec.Command("kubectl", "apply", "-f", "/tmp/autoscaling.yaml")
	response, err = cmd.Output()
	if err != nil {
		log.Println(err)
	} else {
		log.Println("Applied autoscaling template: " + string(response))
	}

	data = []byte(`
apiVersion: v1
kind: ServiceAccount
metadata:
  name: eks-admin
  namespace: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: eks-admin
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
  - kind: ServiceAccount
    name: eks-admin
    namespace: kube-system
`)

	err = os.WriteFile("/tmp/serviceaccount.yaml", data, 0644)
	if err != nil {
		log.Println(err)
	}

	cmd = exec.Command("kubectl", "apply", "-f", "/tmp/serviceaccount.yaml")
	response, err = cmd.Output()
	if err != nil {
		log.Println(err)
	} else {
		log.Println("Applied Service Account template: " + string(response))
	}

	// Download eks charts
	cmd = exec.Command("helm", "repo", "add", "eks", "https://aws.github.io/eks-charts")
	response, err = cmd.Output()
	if err != nil {
		log.Println(err)
	} else {
		log.Println("Installed eks charts:", string(response))
	}

	// install required config
	cmd = exec.Command("kubectl", "apply", "-f", "github.com/aws/eks-charts/stable/aws-load-balancer-controller//crds?ref=master")
	response, err = cmd.Output()
	if err != nil {
		log.Println(err)
	} else {
		log.Println("Installed aws alb controller library: " + string(response))
	}

	// deploy aws alb controller
	cmd = exec.Command("helm", "upgrade", "--install", "--wait", "--timeout", "60s", "--set", "clusterName="+eksClusterName, "--set", "autoDiscoverAwsRegion=true", "--set", "autoDiscoverAwsVpcID=true", "--namespace", "kube-system", "alb-ingress-controller", "eks/aws-load-balancer-controller")
	response, err = cmd.Output()
	if err != nil {
		log.Println(err)
	} else {
		log.Println("Deployed aws alb controller: " + string(response))
	}

	// set docker hub creds
	cmd = exec.Command("kubectl", "create", "secret", "docker-registry", "regcred", "--docker-username="+os.Getenv("DOCKER_USER"), "--docker-password="+os.Getenv("DOCKER_PASS"))
	//response, err = cmd.Output()
	cmd.Stdout = &out
	cmd.Stderr = &stderr
	err = cmd.Run()

	if err != nil {
		log.Println(fmt.Sprint(err) + ": " + stderr.String())
		//log.Println(err)
	} else {
		log.Println("Applied kube template: " + out.String())
	}

	// add solo to library
	cmd = exec.Command("helm", "repo", "add", "gloo", "https://storage.googleapis.com/solo-public-helm")
	response, err = cmd.Output()
	if err != nil {
		log.Println(err)
	} else {
		log.Println("Added gloo library to helm: " + string(response))
	}

	// add solo to library
	cmd = exec.Command("helm", "repo", "update")
	response, err = cmd.Output()
	if err != nil {
		log.Println(err)
	} else {
		log.Println("Deployed aws alb controller: " + string(response))
	}

	// add solo to library
	cmd = exec.Command("helm", "repo", "update")
	response, err = cmd.Output()
	if err != nil {
		log.Println(err)
	} else {
		log.Println("Deployed aws alb controller: " + string(response))
	}

	// get certificate arn
	var certArn string
	s := SECRETS.GetSecrets(cfg, "cms-ssl-certificates", stackName)
	for k, v := range s {
		if k == stackName {
			certArn = v
		}
	}

	gatewayOverrides := []byte(`
settings:
  watchNamespaces:
    - default
gloo:
  logLevel: warn
discovery:
  logLevel: warn
gatewayProxies:
  gatewayProxy:
    gatewaySettings:
      disableHttpsGateway: true
    service:
      httpsFirst: true
      kubeResourceOverride:
        spec:
          ports:
            - name: http
              port: 80
              targetPort: 8080
            - name: https
              port: 443
              targetPort: 8080
      extraAnnotations:
        # https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.2/guide/service/nlb/#configuration
        service.beta.kubernetes.io/aws-load-balancer-type: "external"
        # https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.2/guide/service/nlb/
        service.beta.kubernetes.io/aws-load-balancer-scheme: internet-facing
        service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: "instance"
        service.beta.kubernetes.io/aws-load-balancer-ssl-ports: "443"
        service.beta.kubernetes.io/aws-load-balancer-ssl-cert: ` + certArn + `
        # https://docs.aws.amazon.com/elasticloadbalancing/latest/network/create-tls-listener.html#describe-ssl-policies
        service.beta.kubernetes.io/aws-load-balancer-ssl-negotiation-policy: "ELBSecurityPolicy-TLS13-1-2-2021-06"`)

	err = os.WriteFile("/tmp/gatewayoverrides.yaml", gatewayOverrides, 0644)
	if err != nil {
		log.Println(err)
	}

	cmd = exec.Command("helm", "upgrade", "--install", "--wait", "--namespace", "default", "--values", "/tmp/gatewayoverrides.yaml", "gloo", "gloo/gloo", "--version", "1.10.13")
	log.Println(cmd)
	cmd.Stdout = &out
	cmd.Stderr = &stderr
	err = cmd.Run()

	if err != nil {
		log.Println(fmt.Sprint(err) + ": " + stderr.String())
		//log.Println(err)
	} else {
		log.Println("Deployed gloo api gateway: " + out.String())
	}

	// patch gloo to enable logging
	patch := []byte(`
apiVersion: gateway.solo.io/v1
kind: Gateway
metadata:
  name: gateway-proxy
  namespace: default
spec:
  bindAddress: '::'
  bindPort: 8080
  proxyNames:
    - gateway-proxy
  httpGateway: {}
  useProxyProto: false
  options:
    accessLoggingService:
      accessLog:
      - fileSink:
          jsonFormat:
            httpMethod: '%REQ(:METHOD)%'
            protocol: '%PROTOCOL%'
            responseCode: '%RESPONSE_CODE%'
            clientDuration: '%DURATION%'
            targetDuration: '%RESPONSE_DURATION%'
            path: '%REQ(X-ENVOY-ORIGINAL-PATH?:PATH)%'
            upstreamName: '%UPSTREAM_CLUSTER%'
            systemTime: '%START_TIME%'
            requestId: '%REQ(X-REQUEST-ID)%'
            responseFlags: '%RESPONSE_FLAGS%'
            messageType: '%REQ(x-type)%'
            number: '%REQ(x-number)%'
          path: /dev/stdout`)

	err = os.WriteFile("/tmp/patch.yaml", []byte(patch), 0644)
	if err != nil {
		log.Println(err)
	}

	cmd = exec.Command("kubectl", "apply", "-f", "/tmp/patch.yaml")
	log.Println(cmd)
	cmd.Stdout = &out
	cmd.Stderr = &stderr
	err = cmd.Run()

	if err != nil {
		log.Println(fmt.Sprint(err) + ": " + stderr.String())
		//log.Println(err)
	} else {
		log.Println("Patched gloo api gateway: " + out.String())
	}

	// add datadog to library
	cmd = exec.Command("helm", "repo", "add", "datadog", "https://helm.datadoghq.com")
	response, err = cmd.Output()
	if err != nil {
		log.Println(err)
	} else {
		log.Println("Added datadog repo to helm: " + string(response))
	}

	// get certificate arn
	var ddKey string
	s = SECRETS.GetSecrets(cfg, "datadog", stackName)
	for k, v := range s {
		if k == "datadog_key" {
			ddKey = v
		}
	}

	cmd = exec.Command(
		"helm", "upgrade", "--install", "datadog-agent", "datadog/datadog",
		"--set", "datadog.tags={env:"+stackName+", region:"+vpcStackOutput["Region"]+", cluster="+eksClusterName+"}",
		"--set", "datadog.site=datadoghq.com",
		"--set", "datadog.apiKey="+ddKey,
		"--set", "datadog.clusterName="+stackName,
		"--set", "datadog.containerExclude='image:quay.io/solo-io/gateway image:quay.io/solo-io/discovery image:observeinc/proxy image:gcr.io/datadoghq/agent image:gcr.io/datadoghq/cluster-agent kube_namespace:kube-system kube_namespace:kube-node-lease kube_namespace:kube-public kube_namespace:kubernetes-dashboard'",
		"--set", "datadog.logs.enabled=true",
		"--set", "datadog.apm.portEnabled=true",
		"--set", "datadog.logs.containerCollectAll=true",
		"--set", "targetSystem=linux",
		"--set", "datadog.apm.port=8126",
		"--set", "datadog.serviceTopology.enabled=true")

	response, err = cmd.Output()
	if err != nil {
		log.Println(err)
	} else {
		log.Println("Deployed Datadog to eks cluster: " + string(response))
	}

	// add datadog trace service
	ddtrace := []byte(`
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: datadog-agent
  name: datadog-agent
spec:
  type: ClusterIP
  ports:
    - name: dogstatsd
      port: 8125
      protocol: UDP
      targetPort: 8125
    - name: apm
      port: 8126
      protocol: TCP
      targetPort: 8126
  selector:
    app: datadog-agent
	`)

	err = os.WriteFile("/tmp/ddtrace.yaml", []byte(ddtrace), 0644)
	if err != nil {
		log.Println(err)
	}

	cmd = exec.Command("kubectl", "apply", "-f", "/tmp/ddtrace.yaml")
	response, err = cmd.Output()
	if err != nil {
		log.Println(err)
	} else {
		log.Println("Added dd trace service: " + string(response))
	}

}

File: python-iac/bin/eks-cluster-config.py
#!/usr/bin/python

import argparse
import os
import time

import lib.cloudformation as cfn
import lib.ec2 as ec2
import lib.exec_ctl as exe
import lib.kubeconfig as kube
import lib.params as params
from lib.infrastructure_build_manager import InfrastructureBuildManager

ap = argparse.ArgumentParser()
ap.add_argument("--envName", required=True, help="Required: Environment name equals environment to deploy to")

args = vars(ap.parse_args())

# Create a list of all regions of the env
paramsDir = "./params/" + args["envName"] + "/"

regions_sorted_list = InfrastructureBuildManager.prepare_region_builds(paramsDir, args["envName"])


# Loop through the list of regions and perform the selected action on every region
for region in regions_sorted_list:
    region_name = region["RegionName"]
    # load params into memory
    environment = InfrastructureBuildManager.get_build_environment(region_name, args["envName"], None)
    # temporary fix to stop this script if eks was not deployed
    if environment["DeployEKS"] == "true":
        print("")
        print("Configuring EKS of " + args["envName"] + " env in " + environment["Region"] + " region")
        print("")

        # configure kubeconfig
        clusterName = cfn.get_cluster(args["envName"], environment["Region"])
        kube.update_kubeconfig(clusterName, environment["Region"])

        # tag subnets for alb-ingress-controller automation
        eksClusterName = cfn.get_nested_name(args["envName"], "Eks", environment["Region"])
        stackName = cfn.get_nested_name(args["envName"], "Vpc", environment["Region"])
        PublicSubnetGroup = cfn.get_stack_output(stackName, "PublicSubnetGroup", environment["Region"])
        PrivateSubnetGroup = cfn.get_stack_output(stackName, "PrivateSubnetGroup", environment["Region"])

        # tag public subnets
        for subnet in PublicSubnetGroup.split(","):
            results = ec2.tag_subnet(subnet, "kubernetes.io/cluster/" + eksClusterName, "shared", environment["Region"])
            print(results)
            results = ec2.tag_subnet(subnet, "kubernetes.io/role/elb", "1", environment["Region"])
            print(results)
        # tag private subnets
        for subnet in PrivateSubnetGroup.split(","):
            results = ec2.tag_subnet(subnet, "kubernetes.io/cluster/" + eksClusterName, "shared", environment["Region"])
            print(results)
            results = ec2.tag_subnet(subnet, "kubernetes.io/role/internal-elb", "1", environment["Region"])
            print(results)

        # see if aws-auth has already been added
        eksUser = cfn.get_stack_output(eksClusterName, "EKSUserName", environment["Region"])
        eksUserArn = cfn.get_stack_output(eksClusterName, "EKSUserArn", environment["Region"])
        eksNodeInstanceRole = cfn.get_stack_output(eksClusterName, "EksNodeInstanceRole", environment["Region"])

        awsAuthCheck = "kubectl describe configmap aws-auth -n kube-system"
        status = exe.sub_process(awsAuthCheck)

        awsAuth = (
            """\
apiVersion: v1
kind: ConfigMap
metadata:
  name: aws-auth
  namespace: kube-system
data:
  mapRoles: |
    - rolearn: """
            + eksNodeInstanceRole
            + """
      username: system:node:{{EC2PrivateDNSName}}
      groups:
        - system:bootstrappers
        - system:nodes
    - rolearn: arn:aws:iam::663946581577:role/SAMladmin
      username: SAMladmin
      groups:
        - system:masters
        - cluster-admin
        - tiller
  mapUsers: |
    - userarn: """
            + eksUserArn
            + """
      username: """
            + eksUser
            + """
      groups:
        - system:masters
        - cluster-admin
        - tiller
"""
        )
        authfilepath = "/tmp/" + args["envName"] + "-aws-auth.yaml"
        f = open(authfilepath, "w")
        f.write(awsAuth)
        f.close()

        while os.stat(authfilepath).st_size == 0:
            time.sleep(5)
            print("Waiting for file write out")

        # add aws-auth if doesn't exist already
        if eksUser not in status:
            command = "kubectl apply -f " + authfilepath
            status = exe.sub_process(command)
            print(status)
        else:
            print("aws-auth has the following config:")
            print(status)

        #### setup k8s web console

        # deploy metrics server
        command = "kubectl apply -f https://github.com/kubernetes-sigs/metrics-server/releases/download/v0.3.6/components.yaml"
        status = exe.sub_process(command)
        print(status)

        # Deploy the Kubernetes dashboard

        command = "kubectl apply -f https://raw.githubusercontent.com/kubernetes/dashboard/v2.0.0-beta8/aio/deploy/recommended.yaml"
        status = exe.sub_process(command)
        print(status)

        # Create an eks-admin service account and cluster role binding
        rawClusterAdmin = """\
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: eks-admin
  namespace: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: eks-admin
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
  - kind: ServiceAccount
    name: eks-admin
    namespace: kube-system
"""
        clusterfilepath = "/tmp/" + args["envName"] + "-rawClusterAdmin.yaml"
        f = open(clusterfilepath, "w")
        f.write(rawClusterAdmin)
        f.close()

        command = "kubectl apply -f " + clusterfilepath
        status = exe.sub_process(command)
        print(status)

        # cleanup files
        """if os.path.exists(authfilepath):
          os.remove(authfilepath)
      if os.path.exists(clusterfilepath):
          os.remove(clusterfilepath)"""

        # Read in the autoscaler yaml file
        with open("cluster-autoscaler-autodiscover.yaml") as file:
            autoscalerfiledata = file.read()

        # Replace the Cluster name placeholder string with cluster name
        autoscalerfiledata = autoscalerfiledata.replace("<YOUR_CLUSTER_NAME>", clusterName)

        # Cluster AutoScaler file path per EKS cluster
        autoscalerfilepath = "/tmp/" + clusterName + "-cluster-autoscaler.yaml"

        # Write the file out again
        with open(autoscalerfilepath, "w") as file:
            file.write(autoscalerfiledata)

        # Deploy the Kubernetes Cluster Autoscaler
        command = "kubectl apply -f " + autoscalerfilepath
        status = exe.sub_process(command)
        print(status)

        # Install the nvidia plugin
        command = "kubectl apply -f nvidia-plugin.yaml"
        status = exe.sub_process(command)
        print(status)

        command = f"python3 bin/eks-airflow-cluster-config.py --envName {args['envName']} --setK8Env false"
        status = exe.sub_process(command)
        print(status)

    else:
        print("DeployEKS is false for " + args["envName"] + " env in " + region + " region, so nothing to be done here")
        print("")
